### DISCUSSION TITLE
constraint check failed: EAP identity X required [jumps to the next until finding a match]

### DISCUSSION THREAD

**Discussion by acrolink** [strongswan/strongswan#590] [Q&A] [answered]
I am trying to set up EAP-mschapv2 with different clients, each EAP identity having unique IP address. I have managed to do that using `swanctl.conf` but I see these log messages, maybe I am doing something wrong:

```
charon-systemd[228241]: authentication of 'eap_10_182_10_231' with EAP successful
charon-systemd[228241]: constraint check failed: EAP identity 'eap_10_182_10_6' required
charon-systemd[228241]: selected peer config 'conn-10-182-10-6' unacceptable: non-matching authentication done
charon-systemd[228241]: switching to peer config 'conn-10-182-10-7'
charon-systemd[228241]: constraint check failed: EAP identity 'eap_10_182_10_7' required
charon-systemd[228241]: selected peer config 'conn-10-182-10-7' unacceptable: non-matching authentication done
charon-systemd[228241]: switching to peer config 'conn-10-182-10-8'
```

I this normal?

Here is partial `swantctl.conf`:

```
eap-init {
    mobike = no
    unique = replace
    proposals = aes192gcm16-aes128gcm16-prfsha256-ecp256-ecp521,aes192-sha256-modp3072,default
    include /etc/swanctl/conf.d/ike_sa_default.conf
    fragmentation = yes
    send_certreq=no
    version=2
    remote_addrs=%any
    local_addrs=%any

    children {
        net {
            local_ts=0.0.0.0/0
            esp_proposals = aes192gcm16-aes128gcm16-prfsha256-ecp256-modp3072,aes192-sha256-ecp256-modp3072,default
            start_action=trap
            updown = /usr/local/libexec/ipsec/_updown iptables
        }

    }
    local-0 {
        certs = my_server.pem
        id = @FQDN
    }
}

connections {
    ...
    conn-10-182-10-6 : eap-init {
        vips=10.182.10.6
        pools=pool-subnet-10-182-10-6/32
        remote-10-182-10-6 {
            auth = eap-mschapv2
            eap_id = eap_10_182_10_6

        }
    }

    conn-10-182-10-7 : eap-init {
        vips=10.182.10.7
        pools=pool-subnet-10-182-10-7/32
        remote-10-182-10-7 {
            auth = eap-mschapv2
            eap_id = eap_10_182_10_7

        }
    }

    conn-10-182-10-231 : eap-init {
        vips=10.182.10.231
        pools=pool-subnet-10-182-10-231/32
        remote-10-182-10-231 {
            auth = eap-mschapv2
            eap_id = eap_10_182_10_231

        }
    }
}

pools {
    ...
    pool-subnet-10-182-10-231/32 {
        addrs=10.182.10.231/32
    }

}
```

**Comment by acrolink** [comment, marked as answer]
Seems that the solution is to add `id = eap_10_182_10_x` to remotes.

**Comment by tobiasbrunner** [comment]
> vips=10.182.10.6
> pools=pool-subnet-10-182-10-6/32

This combination makes no sense. Please refer to [VirtualIP](https://wiki.strongswan.org/projects/strongswan/wiki/VirtualIp).

> Seems that the solution is to add `id = eap_10_182_10_x` to remotes.

That won't help, as configuring `eap_id` to a static value currently won't initiate an EAP-Identity exchange. Please refer to [this old issue](https://wiki.strongswan.org/issues/1057) and the linked workarounds (the ipsec.conf one could easily be adapted for swanctl.conf).

**Reply by acrolink** [reply]
> That won't help, as configuring `eap_id` to a static value currently won't initiate an EAP-Identity exchange. Please refer to [this old issue](https://wiki.strongswan.org/issues/1057) and the linked workarounds (the ipsec.conf one could easily be adapted for swanctl.conf).

Matter of fact is that all those `constraint check failed` errors are gone now after I did this:

```
conn-10-182-10-231 : eap-init {
        vips=10.182.10.231
        pools=pool-subnet-10-182-10-231/32
        remote-10-182-10-231 {
            auth = eap-mschapv2
            eap_id = eap_10_182_10_231
            id = eap_10_182_10_231 // added this line.
        }
    }
```
Now it selects the correct connection to authenticate against from first round.

As for `vips`, I will read the documentation to understand better. Thank you.

**Reply by tobiasbrunner** [reply]
Ah, I actually misread that. I thought you configured `eap_id`, not `id` on the clients. `id` (i.e. the IKE Identity as compared to the EAP-Identity/username) is used when initially selecting a config, so there won't be any switches required later.

**Comment by sta8r9** [comment]
这是为什么？
